runtime/src/bank.rs

```diff
@@ -4372,34 +4372,37 @@ impl Bank {
                         feature_set_clone_time.as_us()
                     );
 
-                    let tx_wide_compute_cap = feature_set.is_active(&tx_wide_compute_cap::id());
-                    let compute_budget_max_units = if tx_wide_compute_cap {
-                        compute_budget::MAX_UNITS
+                    let compute_budget = if let Some(compute_budget) = self.compute_budget {
+                        compute_budget
                     } else {
-                        compute_budget::DEFAULT_UNITS
-                    };
-                    let mut compute_budget = self
-                        .compute_budget
-                        .unwrap_or_else(|| ComputeBudget::new(compute_budget_max_units));
-                    if tx_wide_compute_cap {
-                        let mut compute_budget_process_transaction_time =
-                            Measure::start("compute_budget_process_transaction_time");
-                        let process_transaction_result = compute_budget.process_message(
-                            tx.message(),
-                            feature_set.is_active(&requestable_heap_size::id()),
-                            feature_set.is_active(&default_units_per_instruction::id()),
-                        );
-                        compute_budget_process_transaction_time.stop();
-                        saturating_add_assign!(
-                            timings
-                                .execute_accessories
-                                .compute_budget_process_transaction_us,
-                            compute_budget_process_transaction_time.as_us()
-                        );
-                        if let Err(err) = process_transaction_result {
-                            return TransactionExecutionResult::NotExecuted(err);
+                        let tx_wide_compute_cap = feature_set.is_active(&tx_wide_compute_cap::id());
+                        let compute_budget_max_units = if tx_wide_compute_cap {
+                            compute_budget::MAX_UNITS
+                        } else {
+                            compute_budget::DEFAULT_UNITS
+                        };
+                        let mut compute_budget = ComputeBudget::new(compute_budget_max_units);
+                        if tx_wide_compute_cap {
+                            let mut compute_budget_process_transaction_time =
+                                Measure::start("compute_budget_process_transaction_time");
+                            let process_transaction_result = compute_budget.process_message(
```

Comment:
Hmm, I wonder if some of these things -- eg `process_transaction_result`, the `ComputeBudget::process_message` method -- should be renamed. I think that one reason I missed this is that these things seem like things that should happen for all messages being processed.
Maybe this should be something more like:
```suggestion
                            let compute_budget_adjustment_result = compute_budget.adjust(
```
Anyway, not something that needs to be in this PR.

Reply:
Yeah definitely.  I was looking at how to fix this in the compute budget first, but that seems like a different set of work.  It's a bit confusing because there's `MAX_UNITS` (max for whole transaction) and `DEFAULT_UNITS` (max for one instruction) which are both kind of covered as `max_units`.  We might need to separate that field into `max_units_per_instruction` and `max_units_per_transaction`, and then use that when processing the instruction.